include/nbl/builtin/hlsl/bxdf/cook_torrance_base.hlsl

```diff
@@ -17,95 +17,303 @@ namespace hlsl
 namespace bxdf
 {
 
-// N (NDF), F (fresnel), MT (measure transform, using DualMeasureQuant)
-template<class Config, class N, class F, class MT NBL_PRIMARY_REQUIRES(config_concepts::MicrofacetConfiguration<Config> && ndf::NDF<N> && fresnel::Fresnel<F>)
-struct SCookTorrance
+// N (NDF), F (fresnel)
+template<class Config, class N, class F, bool IsBSDF NBL_STRUCT_CONSTRAINABLE>
+struct SCookTorrance;
+
+template<class Config, class N, class F>
+NBL_PARTIAL_REQ_TOP(config_concepts::MicrofacetConfiguration<Config> && ndf::NDF<N> && fresnel::Fresnel<F>)
+struct SCookTorrance<Config, N, F, false NBL_PARTIAL_REQ_BOT(config_concepts::MicrofacetConfiguration<Config> && ndf::NDF<N> && fresnel::Fresnel<F>) >
 {
     NBL_BXDF_CONFIG_ALIAS(scalar_type, Config);
     NBL_BXDF_CONFIG_ALIAS(vector2_type, Config);
     NBL_BXDF_CONFIG_ALIAS(vector3_type, Config);
+    NBL_BXDF_CONFIG_ALIAS(ray_dir_info_type, Config);
     NBL_BXDF_CONFIG_ALIAS(isotropic_interaction_type, Config);
     NBL_BXDF_CONFIG_ALIAS(anisotropic_interaction_type, Config);
     NBL_BXDF_CONFIG_ALIAS(sample_type, Config);
+    NBL_BXDF_CONFIG_ALIAS(spectral_type, Config);
+    NBL_BXDF_CONFIG_ALIAS(quotient_pdf_type, Config);
     NBL_BXDF_CONFIG_ALIAS(isocache_type, Config);
     NBL_BXDF_CONFIG_ALIAS(anisocache_type, Config);
 
-    scalar_type __D(NBL_CONST_REF_ARG(isocache_type) cache)
+    using quant_type = typename N::quant_type;
+
+    spectral_type eval(NBL_CONST_REF_ARG(sample_type) _sample, NBL_CONST_REF_ARG(isotropic_interaction_type) interaction, NBL_CONST_REF_ARG(isocache_type) cache)
     {
-        return ndf.template D<isocache_type>(cache);
+        if (interaction.getNdotV() > numeric_limits<scalar_type>::min)
+        {
+            using quant_query_type = typename N::quant_query_type;
+            using g2g1_query_type = typename N::g2g1_query_type;
+
+            scalar_type dummy;
+            quant_query_type qq = ndf.template createQuantQuery<isocache_type>(cache, dummy);
+            g2g1_query_type gq = ndf.template createG2G1Query<sample_type, isotropic_interaction_type>(_sample, interaction);
```

Comment:
you don't make use of `gq` until the if statement below, then create it there, lowe it into the if